get/DownloadRunnableFallback.java

```diff
@@ -0,0 +1,155 @@
+        } catch (IOException e) {
+            // nothing to do
+        }
+
+        if (mF != null) mF.close();
+    }
+
+    @Override
+    public void run() {
+        boolean done;
+        long start = mMission.fallbackResumeOffset;
+
+        if (DEBUG && !mMission.unknownLength && start > 0) {
+            Log.i(TAG, "Resuming a single-thread download at " + start);
+        }
+
+        try {
+            long rangeStart = (mMission.unknownLength || start < 1) ? -1 : start;
+
+            int mId = 1;
+            mConn = mMission.openConnection(false, rangeStart, -1);
+
+            if (mRetryCount == 0 && rangeStart == -1) {
+                // workaround: bypass android connection pool
+                mConn.setRequestProperty("Range", "bytes=0-");
+            }
+
+            mMission.establishConnection(mId, mConn);
+
+            // check if the download can be resumed
+            if (mConn.getResponseCode() == 416 && start > 0) {
+                mMission.notifyProgress(-start);
+                start = 0;
+                mRetryCount--;
+                throw new DownloadMission.HttpError(416);
+            }
+
+            // secondary check for the file length
+            if (!mMission.unknownLength)
+                mMission.unknownLength = Utility.getContentLength(mConn) == -1;
+
+            if (mMission.unknownLength || mConn.getResponseCode() == 200) {
+                // restart amount of bytes downloaded
+                mMission.done = mMission.offsets[mMission.current] - mMission.offsets[0];
+            }
+
+            mF = mMission.storage.getStream();
+            mF.seek(mMission.offsets[mMission.current] + start);
+
+            mIs = mConn.getInputStream();
+
+            byte[] buf = new byte[DownloadMission.BUFFER_SIZE];
+            int len = 0;
+
+            while (mMission.running && (len = mIs.read(buf, 0, buf.length)) != -1) {
+                mF.write(buf, 0, len);
+                start += len;
+                mMission.notifyProgress(len);
+            }
+
+            dispose();
+
+            // if thread goes interrupted check if the last part is written. This avoid re-download the whole file
+            done = len == -1;
+        } catch (Exception e) {
+            dispose();
+
+            mMission.fallbackResumeOffset = start;
+
+            if (!mMission.running || e instanceof ClosedByInterruptException) return;
+
+            if (e instanceof HttpError && ((HttpError) e).statusCode == ERROR_HTTP_FORBIDDEN) {
+                // for youtube streams. The url has expired, recover
+                dispose();
+                mMission.doRecover(ERROR_HTTP_FORBIDDEN);
+                return;
+            }
+
+            if (mRetryCount++ >= mMission.maxRetry) {
+                mMission.notifyError(e);
+                return;
+            }
+
+            if (DEBUG) {
+                Log.e(TAG, "got exception, retrying...", e);
+            }
+
+            run();// try again
```

Comment:
<div>


<div id="suggestion">
<div id="issue"><b>Consider refactoring retry mechanism implementation</b></div>
<div id="fix">

Consider extracting the retry logic into a separate method to improve code readability and maintainability. The recursive call to `run()` could lead to stack overflow for many retries.
</div>
<details>
<summary>
<b>Code suggestion</b>
</summary>
<div id="code">

```
 @@ -52,83 +52,89 @@
     public void run() {
         boolean done;
         long start = mMission.fallbackResumeOffset;
 +       while(mRetryCount <= mMission.maxRetry && mMission.running) {
             try {
                 // download logic
                 if (done) {
                     mMission.notifyFinished();
                     break;
                 }
             } catch (Exception e) {
                 if (!mMission.running || e instanceof ClosedByInterruptException) return;
                
                 if (e instanceof HttpError && ((HttpError) e).statusCode == ERROR_HTTP_FORBIDDEN) {
                     dispose();
                     mMission.doRecover(ERROR_HTTP_FORBIDDEN);
                     return;
                 }

                 if (mRetryCount++ >= mMission.maxRetry) {
                     mMission.notifyError(e);
                     return;
                 }

                 if (DEBUG) {
                     Log.e(TAG, "got exception, retrying...", e);
                 }
                 continue;
             }
         }
```
</div>
</details>
</div>


<small><i>Code Review Run <a href=https://github.com/OBrutus/youtube-downloader/pull/9#issuecomment-2464377435>#3b9ab6</a></i></small>
</div>

---
Is this a valid issue, or was it incorrectly flagged by the Agent?
- [ ] it was incorrectly flagged